### PR TITLE
(DOCSP-29203): Aggregates Builders page

### PR DESCRIPTION
**Pull request by cbullinger**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-java/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-29203
Staging - https://docs-mongodbcom-staging.corp.mongodb.com/kotlin/docsworker-xlarge/docsp-29203-aggregates/fundamentals/builders/aggregates/

## Updated Page

Fundamentals > Builders > [Aggregates Builder page](https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/builders/aggregates/)

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST?
- [ ] Did you run a spell-check?
- [ ] Did you run a grammar-check?
- [ ] Are all the links working?

### CODE REVIEW

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 1]
```diff
@@ -0,0 +1,1028 @@
...
+            // aggregate content to merge
+            Aggregates.group("\$${Movie::year.name}", Accumulators.bottom(Results::lowestRatedTwoMovies.name, Sorts.descending("${Movie::imdb.name}.${Movie.IMDB::rating.name}"), listOf("\$${Movie::title.name}", "\$${Movie::imdb.name}.${Movie.IMDB::rating.name}"))),
+                // :snippet-start: merge
+                Aggregates.merge("comments")
+                // :snippet-end:
+            ))
+        assertEquals(5, resultsFlow.toList().size)
+    }
+
+//    @Test
```
[the remark this answers is not in the document]
Comment (mongodben) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt]: fixed

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 2]
```diff
@@ -0,0 +1,1005 @@
...
+    @Test
+    fun mergeOptionsTest() = runBlocking {
+       // movieCollection.createIndex(Indexes.ascending("year", "title"))
+
+        val resultsFlow = movieCollection.aggregate<Movie>(listOf(
+            Aggregates.project(fields(Projections.computed("year", "\$_id - \$year"), Projections.include("title", "year"))),
+                // :snippet-start: merge-options
+                Aggregates.merge(
+                    MongoNamespace("aggregation", "movies"), // ERROR: 'Cannot find index to verify that join fields will be unique'
+                    MergeOptions().uniqueIdentifier(listOf("year", "title"))
```
[the remark this answers is not in the document]
Comment (mongodben) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt:619]: fixed

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 3]
```diff
@@ -0,0 +1,1005 @@
...
+        ))
+        resultsFlow.collect{println(it)}
+
+        weatherCollection.drop()
+    }
+
+    @Test
+    fun fullTextSearchTest(): Unit = runBlocking {
+        ftsCollection.insertOne(Movie(id = 1, title = "Back to the Future", year = 1985, cast = listOf("Michael J. Fox", "Christopher Lloyd"), genres = listOf("Adventure", "Comedy", "Sci-Fi"), type = "movie", rated = "PG", plot = "Marty McFly, a 17-year-old high school student, is accidentally sent thirty years into the past in a time-traveling DeLorean.", fullplot = "Marty McFly, a 17-year-old high school student, is accidentally sent thirty years into the past in a time-traveling DeLorean invented by his close friend, the eccentric scientist Doc Brown.", runtime = 116, imdb = Movie.IMDB(rating = 8.5)))
+        // the Atlas search index is enabled on ftsCollection: db ("sample_mflix") and collection ("movies")
```
[the remark this answers is not in the document]
Comment (mongodben) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt]: fixed

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 4]
```diff
@@ -0,0 +1,947 @@
...
+        val orders = listOf(Order(1, 1, "item1", 5), Order(2, 1, "item2", 5), Order(3, 1, "item3", 5))
+        val inventory= listOf(Inventory(1, "item1", 10), Inventory(2, "item2", 20), Inventory(3, "item3", 30))
+
+        orderCollection.insertMany(orders)
+        warehouseCollection.insertMany(inventory)
+        // TODO: Confirm why using Filters doesn't populate the stockData field? aren't these equivalent?
+        // Filters.and(
+        //  Filters.eq("order_item", Inventory::stockItem.name),
+        //  Filters.gte(Inventory::inStock.name, "order_qty")
+        //)
```
Comment (mongodben) [inline examples/src/test/kotlin/AggregatesBuilderTest.kt:291]: i also couldn't get this example to work with builders. though here's what i think this should be (slightly different from yours):

```kt
Filters.and(
  Filters.eq("\$\$order_item", "\$${Inventory::stockItem.name}"),
  Filters.gte(Inventory::inStock.name, "\$\$order_qty")
)
```

however, using this gives me the following error:

```
Command failed with error 168 (InvalidPipelineOperator): 'Unrecognized expression '$$order_item'' on server ac-j8rabip-shard-00-01.nzumlgv.mongodb.net:27017. The full response is {"ok": 0.0, "errmsg": "Unrecognized expression '$$order_item'", "code": 168, "codeName": "InvalidPipelineOperator", "$clusterTime": {"clusterTime": {"$timestamp": {"t": 1684960589, "i": 27}}, "signature": {"hash": {"$binary": {"base64": "LYBjwfW57eKTP5PI8+nGLspt4Sk=", "subType": "00"}}, "keyId": 7200462029784612865}}, "operationTime": {"$timestamp": {"t": 1684960589, "i": 27}}}
com.mongodb.MongoCommandException: Command failed with error 168 (InvalidPipelineOperator): 'Unrecognized expression '$$order_item'' on server ac-j8rabip-shard-00-01.nzumlgv.mongodb.net:27017. The full response is {"ok": 0.0, "errmsg": "Unrecognized expression '$$order_item'", "code": 168, "codeName": "InvalidPipelineOperator", "$clusterTime": {"clusterTime": {"$timestamp": {"t": 1684960589, "i": 27}}, "signature": {"hash": {"$binary": {"base64": "LYBjwfW57eKTP5PI8+nGLspt4Sk=", "subType": "00"}}, "keyId": 7200462029784612865}}, "operationTime": {"$timestamp": {"t": 1684960589, "i": 27}}}
	at app//com.mongodb.internal.connection.ProtocolHelper.getCommandFailureException(ProtocolHelper.java:205)
	at app//com.mongodb.internal.connection.InternalStreamConnection.lambda$sendCommandMessageAsync$0(InternalStreamConnection.java:552)
	at app//com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback$MessageCallback.onResult(InternalStreamConnection.java:858)
	at app//com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback$MessageCallback.onResult(InternalStreamConnection.java:821)
	at app//com.mongodb.internal.connection.InternalStreamConnection$3.completed(InternalStreamConnection.java:680)
	at app//com.mongodb.internal.connection.InternalStreamConnection$3.completed(InternalStreamConnection.java:677)
	at app//com.mongodb.internal.connection.AsynchronousChannelStream$BasicCompletionHandler.completed(AsynchronousChannelStream.java:251)
	at app//com.mongodb.internal.connection.AsynchronousChannelStream$BasicCompletionHandler.completed(AsynchronousChannelStream.java:234)
	at app//com.mongodb.internal.connection.tlschannel.async.AsynchronousTlsChannel.lambda$read$4(AsynchronousTlsChannel.java:122)
	at java.base@11.0.12/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base@11.0.12/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base@11.0.12/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base@11.0.12/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base@11.0.12/java.lang.Thread.run(Thread.java:829)
```

curious for tech reviewer's feedback here

Comment (mongodben) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt:291]: for now, i think we can just use the Document-based version

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 5]
```diff
@@ -0,0 +1,1028 @@
...
+                // :snippet-end:
+            )
+        )
+        assertEquals(5, resultsFlow.toList().size)
+    }
+
+    @Test
+    fun topTest() = runBlocking {
+        data class TopRated(val title: String, val rating: Movie.IMDB)
+        data class Results(@BsonId val id: Int, val topRatedMovie: List<TopRated>) // TODO ERROR: Unable to decode topRatedMovie for Results data class (Document{{_id=1972, topRatedMovie=[The Godfather, 8.9]}})
```
[the remark this answers is not in the document]
Comment (mongodben) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt]: yeah this is tricky b/c the result includes a mixed array:

```json
{"_id": 2008, "topRatedMovie": ["The Dark Knight", 5.2]}

```

this doesn't play nice with kotlin data types. using Document class is prob easiest. 

as a workaround, i tried to do results data class as:

```kt
data class Results(@BsonId val id: Int, val topRatedMovie: List<Any>)
```

but this gives a codec error: 

```
Can't find a codec for CodecCacheKey{clazz=class java.lang.Object, types=null}.
org.bson.codecs.configuration.CodecConfigurationException: Can't find a codec for CodecCacheKey{clazz=class java.lang.Object, types=null}.
	at app//org.bson.internal.ProvidersCodecRegistry.lambda$get$0(ProvidersCodecRegistry.java:84)
	at java.base@11.0.12/java.util.Optional.orElseGet(Optional.java:369)
	at app//org.bson.internal.ProvidersCodecRegistry.get(ProvidersCodecRegistry.java:74)
	at app//org.bson.internal.ChildCodecRegistry.get(ChildCodecRegistry.java:67)
	at app//org.bson.codecs.ContainerCodecHelper.getCodec(ContainerCodecHelper.java:71)
	at app//org.bson.codecs.CollectionCodec.parameterize(CollectionCodec.java:87)
	at app//org.bson.internal.ProvidersCodecRegistry.lambda$get$0(ProvidersCodecRegistry.java:79)
	at java.base@11.0.12/java.util.Optional.orElseGet(Optional.java:369)
...
```

so i think just using the `Document` type is fine here

[the remark this answers is not in the document]
Comment (mongodben) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt]: worth checking if same applies for other topN examples

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 6]
```diff
@@ -0,0 +1,1028 @@
...
+            // :snippet-end:
+        ))
+        assertEquals(4, resultsFlow.toList().size)
+    }
+
+    @Test
+    fun bucketAutoTest() = runBlocking {
+        data class MinMax(val min: Int, val max: Int)
+        data class Results(@BsonId val id: List<MinMax>, val count: Int)
+        val resultsFlow = screenCollection.aggregate<Document>(listOf( // TODO: figure out data class (Document{{_id=Document{{min=15, max=27}}, count=1}})
```
[the remark this answers is not in the document]
Comment (mongodben) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt]: fixed

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 7]
```diff
@@ -0,0 +1,1028 @@
...
+             // :snippet-start: match
+            Aggregates.match(Filters.eq(Movie::title.name, "The Shawshank Redemption"))
+             // :snippet-end:
+        ))
+        assertEquals("The Shawshank Redemption", match.toList().first().title)
+    }
+
+    @Test
+    fun projectTest()  = runBlocking {
+        data class Results(val title: String, val plot: String)
```
[the remark this answers is not in the document]
Comment (mongodben) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt:140]: i would keep as close to what's in java docs

**Review by mongodben — CHANGES_REQUESTED** [review]
updated the code so that everything is passing. i couldn't provide full answers to all your questions, but i believe i responded to everything where i'm not sure the best approach

File: examples/src/test/kotlin/AggregatesBuilderTest.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbullinger) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt:620]: `IndexOptions().unique(true)` 🤦‍♀️

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 8]
```diff
@@ -0,0 +1,939 @@
...
+            Comment(id = ObjectId(), name = "Andrea Le", movie_id = 1, text = "Rem officiis eaque repellendus amet eos doloribus."),
+            Comment(id = ObjectId(), name = "Jane Doe", movie_id = 2, text = "I didn't like this movie. It was boring and predictable."),
+            Comment(id = ObjectId(), name = "Bob Smith", movie_id = 3, text = "This movie is a masterpiece. Highly recommended."))
+        commentCollection.insertMany(comments)
+
+        val lookup = movieCollection.aggregate<Document>(listOf(
+            // :snippet-start: lookup
+            Aggregates.lookup("comments", "_id", Comment::movie_id.name, Results::joined_comments.name
+            // :snippet-end:
+        )))
```
Comment (mongodben) [inline examples/src/test/kotlin/AggregatesBuilderTest.kt:256]: b/c `Comment` and `Results` aren't in the example, think makes more sense to use string literals. also closing paren isn't in the snippet now:

```suggestion
            Aggregates.lookup("comments", "_id", "movie_id.name", "joined_comments")
            // :snippet-end:
        ))
```

Comment (mongodben) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt:256]: might also wanna wrap the snippet across multiple lines for readability

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 9]
```diff
@@ -0,0 +1,939 @@
...
+        orderCollection.insertMany(orders)
+
+        data class Results(val totalQuantity: Int, val averageQuantity: Double)
+
+        val grouping = orderCollection.aggregate<Results>(listOf(
+            // :snippet-start: group
+            Aggregates.group("\$${Order::customerId.name}",
+                sum(Results::totalQuantity.name, "\$${Order::ordered.name}"),
+                avg(Results::averageQuantity.name, "\$${Order::ordered.name}")
+            )
```
Comment (mongodben) [inline examples/src/test/kotlin/AggregatesBuilderTest.kt:346]: again includes results data class w/o defining it. pls replace w string literals.

same comment applies to any other instances of this pattern

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 10]
```diff
@@ -0,0 +1,939 @@
...
+
+    @Test
+    fun graphLookupTest() = runBlocking {
+        data class Results(val name: String, val friends: List<String>, val socialNetwork: List<Users>)
+        val resultsFlow = contactsCollection.aggregate<Results>(listOf(
+                // :snippet-start: graph-lookup
+                Aggregates.graphLookup(
+                    "contacts",
+                    "\$${Users::friends.name}", Users::friends.name, Users::name.name, Results::socialNetwork.name
+                )
```
Comment (mongodben) [inline examples/src/test/kotlin/AggregatesBuilderTest.kt:628]: wrap across more lines

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 11]
```diff
@@ -0,0 +1,939 @@
...
+
+    @Test
+    fun graphLookupDepthTest() = runBlocking {
+        data class Depth(val name: String, val degrees: Int)
+        data class Results(val name: String, val friends: List<String>, val socialNetwork: List<Depth>)
+        val resultsFlow = contactsCollection.aggregate<Results>(listOf(
+                // :snippet-start: graph-lookup-depth
+                Aggregates.graphLookup(
+                    "contacts",
+                    "\$${Users::friends.name}", Users::friends.name, Users::name.name, Results::socialNetwork.name,
```
Comment (mongodben) [inline examples/src/test/kotlin/AggregatesBuilderTest.kt:642]: wrap across more lines

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 12]
```diff
@@ -0,0 +1,939 @@
...
+            Aggregates.sort(Sorts.descending(Results::count.name, "_id"))))
+        val results = resultsFlow.toList()
+        val actual = listOf(Results("Drama", 8), Results("Crime", 3), Results("Action", 2), Results("Thriller", 1), Results("Sci-Fi", 1), Results("Romance", 1), Results("Mystery", 1),)
+        assertEquals(results, actual)
+    }
+
+    @Test
+    fun replaceRootTest() = runBlocking {
+        // :snippet-start: replace-root-data-class
+        data class Libro(val titulo: String)
```
Comment (mongodben) [inline examples/src/test/kotlin/AggregatesBuilderTest.kt:687]: muy bien

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 13]
```diff
@@ -0,0 +1,939 @@
...
+        translateCollection.drop()
+    }
+
+    @Test
+    fun addFieldsTest() = runBlocking {
+        data class Results (val cast: List<String>, val type: String)
+        val resultsFlow = movieCollection.aggregate<Results>(listOf(
+            // :snippet-start: add-fields
+            addFields(Field(
+                Results::cast.name, listOf("Nicholas Cage")),
```
Comment (mongodben) [inline examples/src/test/kotlin/AggregatesBuilderTest.kt:711]: modest proposal to only use Nicholas Cage movies when doing examples with the mflix data  set?

File: source/fundamentals/builders/aggregates.txt [hunk 14]
```diff
@@ -700,15 +635,17 @@ Use the ``bucket()`` method to create a :manual:`$bucket </reference/operator/ag
 pipeline stage that automates the bucketing of data around predefined boundary
 values.
 
-The following example creates a pipeline stage that groups incoming documents based
+The following examples use data modeled with the following Kotlin data class:
+
+.. literalinclude:: /examples/generated/AggregatesBuilderTest.snippet.bucket-data-class.kt
```
Comment (mongodben) [inline source/fundamentals/builders/aggregates.txt:640]: a few sections use this example even though it's only defined in this section. can you redefine it in other sections where it's used?

**Review by mongodben — CHANGES_REQUESTED** [review]
back to you! looks pretty good, just some small things throughout

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 15]
```diff
@@ -0,0 +1,939 @@
...
+
+    @Test
+    fun projectComputeTest()  = runBlocking {
+        data class Results(val rating: String)
+        val compute = movieCollection.aggregate<Results>(
+            listOf(
+                // :snippet-start: project-computed
+                Aggregates.project(
+                    Projections.fields(
+                        Projections.computed(Results::rating.name, "\$${Movie::rated.name}"),
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt:162]: Updated

File: examples/src/test/kotlin/AggregatesBuilderTest.kt (hunk 8 above)
Comment (cbullinger) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt:256]: Updated

File: examples/src/test/kotlin/AggregatesBuilderTest.kt (hunk 9 above)
Comment (cbullinger) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt:346]: Updated throughout

File: source/fundamentals/builders/aggregates.txt (hunk 14 above)
Comment (cbullinger) [reply source/fundamentals/builders/aggregates.txt:640]: Added

File: examples/src/test/kotlin/AggregatesBuilderTest.kt (hunk 11 above)
Comment (cbullinger) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt]: Updated

File: examples/src/test/kotlin/AggregatesBuilderTest.kt (hunk 10 above)
Comment (cbullinger) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt]: Updated

File: examples/src/test/kotlin/AggregatesBuilderTest.kt [hunk 16]
```diff
@@ -0,0 +1,939 @@
...
+        val movieRatings = database.getCollection<Document>("movie_ratings")
+        movieRatings.createIndex(Indexes.ascending("year", "title"), uniqueIndexOption)
+
+        val resultsFlow = movieCollection.aggregate<Document>(listOf(
+            Aggregates.match(Filters.eq(Movie::title.name, "The Sixth Sense")),
+            Aggregates.project(Projections.fields(Projections.computed("rating", "\$imdb.rating"), Projections.include("title", "year"))),
+            // :snippet-start: merge-options
+            Aggregates.merge(
+                MongoNamespace("aggregation", "movie_ratings"),
+                MergeOptions().uniqueIdentifier(listOf("year", "title"))
```
Comment (mongodben) [inline examples/src/test/kotlin/AggregatesBuilderTest.kt:592]: should this be references to property names, like:

```suggestion
                MergeOptions().uniqueIdentifier(listOf(Movie::year.name, Movie::title.name))
```

File: examples/src/test/kotlin/AggregatesBuilderTest.kt (hunk 4 above)
Comment (rozza) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt]: Yeah this is a complex one. Firstly, have to you `$expr` operator to access the variables from the lookup. Secondly `Filters.eq` & `Filters.gte` don't output the correct MQL for use with `$expr` :(

**Review by rozza — APPROVED** [review]
LGTM!

**Review by rozza — APPROVED** [review]
LGTM!

File: examples/src/test/kotlin/AggregatesBuilderTest.kt (hunk 4 above)
Comment (cbullinger) [reply examples/src/test/kotlin/AggregatesBuilderTest.kt]: 👍 thanks for the explanation. I'll remove the TODO

**Pull request merged**